address.go

```diff
@@ -81,6 +118,37 @@ type Address interface {
 // When the address does not encode the network, such as in the case of a raw
 // public key, the address will be associated with the passed defaultNet.
 func DecodeAddress(addr string, defaultNet *chaincfg.Params) (Address, error) {
+
+	// Bech32 encoded segwit addresses start with an human-readable part
+	// (hrp) followed by 1. For Bitcoin mainnet the hrp is "bc", and for
+	// testnet it is "tb". If the address string matches the hrp found in
+	// the defaultNet params, we try to decode it as a segwit address.
+	hrp := defaultNet.Bech32HRPSegwit
+	prefix := hrp + "1"
+	if strings.HasPrefix(strings.ToLower(addr), prefix) {
```

Comment:
Correct.  Although you'll also of course need to make sure you're not using an invalid index as the above would do.  It would panic in the case of a malformed address.

```
	// The network for a bech32 is identified by a human-readable part followed
	// by a 1.
	oneIndex := strings.LastIndexByte(addr, '1')
	if oneIndex != -1 {
			// ....
	}
```